rfcs/git-sdk-requirements/RFC.md

```diff
@@ -0,0 +1,697 @@
+
+There is documented [API on calling WASM](https://developer.mozilla.org/en-US/docs/WebAssembly/Using_the_JavaScript_API).
+
+Need to read through the API exposed by libgit2 to say for sure.
+
+> note: not sure what the interface of FS that libgit2 expects, need to test
+
+#### Q: Do you need to run Git operations in Git SDK in a web worker?
+
+Assuming we figure out how TS communicates nicely with Git commands that get executed in WASM. And for changes to reflect in a file system provided, there would ne no need for a web worker to be there.
+
+#### Q: Can we bundle WASM in Git SDK and abstract using WASM over nice API?
+
+[rollup/plugin-wasm](https://www.npmjs.com/package/@rollup/plugin-wasm) can potentially be used.
+
+However when it was used last time, there were errors.
+
+#### Q: Will using libgit2 compiled to WASM answer to all stated goals above?
+
+✅ = already done
+🚧 = work required
+❎ = not possible
+
+Goal 1: Must run in the browser/on the client [High Confidence] ✅
+
+Calling into `.wasm` from JS, then in turn saving everything in the FS provided.
+
+Goal 2: Lazy loading of files and git history [High Confidence] 🚧 close to ✅ potentially
+
+sparse checkout is supported on a fork for libgit2. It needs to be compiled and tested.
+
+JS code needs to be written to abstract the lazy loading using sparse checkout to make it possible.
+
+Goal 3: Must be git compatible but not necessarily up to spec [Medium Confidence] ✅
+
+No issue here for same reason as Isomorphic Git. We use commands from libgit2 we need.
+
+However the barrier to contributing to libgit2 itself is going to be that command is fully spec compliant so we need to be aware of that.
+
+Goal 4: (Future) File-based auth [High Confidence | Server-related ] 🚧
+
+Is part of server so outside of scope.
+
+> note: maybe wrong
+
+Goal 5: (Future) Support for large files out of the box [Medium Confidence] 🚧/✅
+
+There is open discussion [here](https://github.com/git-lfs/git-lfs/issues/375). But from looks of it, it's supported but may bring issues.
+
+Goal 6: (Future) Plugin system to support different files [Medium Confidence] 🚧
+
+Would be doable to add irrespective of whether libgit2 or Isomorphic Git is chosen.
+
+## Conclusion
+
+Having written above, I believe going with libgit2 compiled to WASM option will allow us to potentially move faster in near future. As we won't need to implement sparse-checkout and rebase.
+
+Both features need some git know how and time to implement.
+
+With libgit2 we get those features out of the box.
+
+Performance is not a concern in either case so the other focus is on DX.
+
+I think you can provide the same Git SDK API regardless whether Isomorphic Git is chosen or libgit2 to do git operations.
+
+As for regretting choose either technology in the future. We would potentially need to think through different kinds of applications that will be built with Git SDK. Some are done as part of this RFC below.
```

Comment:
> Having written above, I believe going with libgit2 compiled to WASM option will allow us to potentially move faster in near future. **As we won't need to implement sparse-checkout and rebase.** 

This conclusion is missing the goal of this RFC. This RFC is not about libgit2 has more features. This RFC is about choosing the right architecture for the next years. Implementing sparse-checkout and rebase in JS shouldn't be an issue.